File: src/shared/config/task.ts
import dayjs from "dayjs";

import type { TaskActionKind, TaskStatus } from "@/shared/domain/task";

/** "view" 纯前端概念（打开详情弹窗），不需要后端参与，不在 TaskActionKind 里。 */
export type TaskAction = TaskActionKind | "view";
type TaskTone = "completed" | "failed" | "queued" | "running";

type TaskActionPresentation = {
  action: TaskAction;
  actionLabel: string;
};

type TaskPresentation = {
  actions: TaskActionPresentation[];
  label: string;
  stage: string;
  tone: TaskTone;
};

export const taskStatusPresentation: Record<TaskStatus, TaskPresentation> = {
  canceled: {
    actions: [{ action: "retry", actionLabel: "重试" }],
    label: "已取消",
    stage: "已取消",
    tone: "completed",
  },
  failed: {
    actions: [{ action: "retry", actionLabel: "重试" }],
    label: "失败",
    stage: "已中断",
    tone: "failed",
  },
  paused: {
    actions: [
      { action: "resume", actionLabel: "继续" },
      { action: "restart", actionLabel: "重新开始" },
    ],
    label: "已暂停",
    stage: "已暂停",
    tone: "queued",
  },
  pending: {
    actions: [
      { action: "cancel", actionLabel: "取消" },
      { action: "pause", actionLabel: "暂停" },
    ],
    label: "排队中",
    stage: "等待调度",
    tone: "queued",
  },
  running: {
    actions: [{ action: "view", actionLabel: "查看" }],
    label: "运行中",
    stage: "执行中",
    tone: "running",
  },
  success: {
    actions: [{ action: "view", actionLabel: "查看结果" }],
    label: "已完成",
    stage: "已完成",
    tone: "completed",
  },
  interrupted: {
    actions: [{ action: "retry", actionLabel: "重试" }],
    label: "已中断",
    stage: "已中断",
    tone: "failed",
  },
};

export const formatTaskTime = (value?: string | null): string => {
  if (!value) return "—";
  const date = dayjs(value);
  return date.isValid() ? date.format("YYYY-MM-DD HH:mm:ss") : value;
};

File: src-tauri/src/domain/task.rs
use crate::domain::module::ModuleId;
use rusqlite::types::{FromSql, FromSqlError, FromSqlResult, ToSql, ToSqlOutput, ValueRef};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use ts_rs::TS;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, TS)]
#[serde(rename_all = "lowercase")]
#[ts(export, export_to = "task.ts")]
pub enum TaskStatus {
    Pending,
    Running,
    Paused,
    Success,
    Failed,
    Canceled,
    Interrupted,
}

impl TaskStatus {
    /// 与数据库 status 列的取值保持一致
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Pending => "pending",
            Self::Running => "running",
            Self::Paused => "paused",
            Self::Success => "success",
            Self::Failed => "failed",
            Self::Canceled => "canceled",
            Self::Interrupted => "interrupted",
        }
    }
}

impl ToSql for TaskStatus {
    fn to_sql(&self) -> rusqlite::Result<ToSqlOutput<'_>> {
        Ok(ToSqlOutput::from(self.as_str()))
    }
}

impl FromSql for TaskStatus {
    fn column_result(value: ValueRef<'_>) -> FromSqlResult<Self> {
        match value.as_str()? {
            "pending" => Ok(Self::Pending),
            "running" => Ok(Self::Running),
            "paused" => Ok(Self::Paused),
            "success" => Ok(Self::Success),
            "failed" => Ok(Self::Failed),
            "canceled" => Ok(Self::Canceled),
            "interrupted" => Ok(Self::Interrupted),
            _ => Err(FromSqlError::InvalidType),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export, export_to = "task.ts")]
pub struct Task {
    #[ts(type = "number")]
    pub id: i64,
    pub task_uid: String,
    pub module_id: ModuleId,
    pub manifest_key: String,
    pub title: String,
    pub params_json: String,
    pub result_json: Option<String>,
    pub status: TaskStatus,
    pub progress: u8,
    #[ts(type = "number")]
    pub total: i64,
    #[ts(type = "number")]
    pub done: i64,
    pub error_message: Option<String>,
    pub created_at: String,
    pub updated_at: String,
    pub started_at: Option<String>,
    pub finished_at: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export, export_to = "task.ts")]
pub struct NewTask {
    pub module_id: ModuleId,
    pub manifest_key: String,
    #[ts(type = "Record<string, unknown>")]
    pub params: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export, export_to = "task.ts")]
pub struct CreateTaskResult {
    pub task_uid: String,
}

/// 队列侧的状态操作——都只是改数据库里的记录，不涉及真正终止/唤起一个执行中的进程
/// （执行引擎落地前，任务永远不会真的处于"有进程在跑"的中间态）。
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, TS)]
#[serde(rename_all = "lowercase")]
#[ts(export, export_to = "task.ts")]
pub enum TaskActionKind {
    Cancel,
    Pause,
    Resume,
    Retry,
    Restart,
}

impl TaskActionKind {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Cancel => "cancel",
            Self::Pause => "pause",
            Self::Resume => "resume",
            Self::Retry => "retry",
            Self::Restart => "restart",
        }
    }
}
